File: WC3/C++_homework/Pen&Paper/Pen.cpp
#include "Pen.h"

Pen::Pen(int inkCapacity) {
    this->inkAmount = 0;
    this->inkCapacity = inkCapacity;
}
Pen::~Pen() {}

int Pen::getInkAmount() const {
    return this->inkAmount;
}
int Pen::getInkCapacity() const {
    return this->inkCapacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    int length = message.size();
    int paperCapacity = paper.getMaxSymbols() - paper.getSymbols();

    if ( this->inkAmount >= length && paperCapacity >= length ) {
        this->inkAmount -= length;
        paper.addContent(message);
        return;
    } else if ( this->inkAmount > paperCapacity ) {
        std::string cutMessage = message.substr(0, this->inkAmount);

        this->inkAmount -= paperCapacity;
        paper.addContent(cutMessage);
        return;
    } else if ( this->inkAmount < paperCapacity ) {
        std::string cutMessage = message.substr(0,this->inkAmount);

        this->inkAmount = 0;
        paper.addContent(cutMessage);
        throw OutOfInk();
    }
    this->inkAmount = 0;
    paper.addContent(message);
    std::cout << "NOPE" << std::endl;
    throw OutOfInk();
}

void Pen::refill() {
    this->inkAmount = this->inkCapacity;
}

File: WC4/PHP/Pen&Paper/Pen.php
<?php
require_once('Exceptions.php');

    class Pen {
        private $inkAmount;
        private $inkCapacity;
        
        private function validate($value) {
            if ( is_int($value) && $value > 0 ) {
                return $value;
            }
            throw new Exception("Invalid param");
        }
        
        public function __construct($inkCapacity = 4096) {
            $this->inkCapacity = $this->validate($inkCapacity);
            $this->inkAmount = $inkCapacity;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new Exception("Error: attribute $name doesn't exist");
        }
        
        public function write(Paper $paper, $message) {
            if ( $this->inkAmount == 0 ) {
                throw new OutOfInk("Action error: no ink left");
            }
            
            $spaceLeft = $paper->maxSymbols - $paper->symbols;
            
            if ( strlen($message) > $this->inkAmount && strlen($message) > $spaceLeft ) {
                if ( $this->inkAmount >= $spaceLeft ) {
                    $this->inkAmount -= $spaceLeft;
                    
                    $paper->addContent(substr($message, 0, $spaceLeft));
                } else {
                    $paper->addContent(substr($message, 0, $this->inkAmount));
                    $this->inkAmount = 0;
                    throw new Exception("Action error: no ink left");
                }
            }
            $paper->addContent($message);
            $this->inkAmount -= $paper->symbols - $beforePaperSymbols;
        }
        
        public function refill() {
            $this->inkAmount = $this->inkCapacity;
        }
    }
?>